CAS-RCH Integration

mkangia opened the pull request:
[Spec](https://docs.google.com/document/d/1BKIGss_R1HRERSW6uHtuJ5GMYIF2hj4eBbYtAD6o-T4/edit#)

mkangia committed:
Add RCH integration to pull data from RCH API and list out [skip ci]

mkangia committed:
extract a helper method for field choices

mkangia committed:
extract updating beneficiaries to parent class instead

mkangia committed:
use get to fetch value

mkangia committed:
require super access to view beneficiaries and beneficiary details

mkangia committed:
add periodic tasks for fetching beneficiaries at 18:00 and 18:15

mkangia committed:
remove unnecessary settings as custom.rch already in HQ_APPS

mkangia committed:
Add MDDS_VillageID to RCHRecord instead of just RCHChild

mkangia committed:
fix time for periodic task to be in PM [skip ci]

mkangia committed:
filter using MDDS_VillageID instead of Village_ID

mkangia committed:
fix dimagimon complaints [skip ci]

mkangia committed:
Add clear button that clears filter and ugettext_lazy -> _

mkangia committed:
move filter submit as a GET request and include total records in the view as well until pagination integrated

Move filter submit to GET to have params in url and avoid form submission request on browser when navigating back after searching and clicking for beneficiary's details

mkangia committed:
Include an empty column for CAS fields

This would help compare values between two systems, since CAS is not available just include N/A

mkangia committed:
update to use GET for filtering by village ids when included

mkangia committed:
minor styling improvements [skip ci]

mkangia committed:
Add email notifiers for Unavailable service, insecure access response and parsing error

mkangia committed:
Rescue and notify in case Service Unavailable, Response Parsing error or insecure access in case of no records [skip ci]

mkangia committed:
update existing beneficiaries using Registration as pk for RCHMother and Child_RCH_ID_No for RCHChild [skip ci]

mkangia committed:
fix dimagimon indentation complaints

mkangia committed:
Fetch Credentials from settings and raise Error in case request initiated without adding them [skip ci]

mkangia committed:
update acceptable fields to be in son for usability, corrected couple of names and added missing fields

mkangia committed:
sanitise and accept only allowed fields, allow to pass number of days prior data to be fetched [skip ci]

mkangia committed:
update village id, name fields [skip ci]

mkangia committed:
add all fields being received from api

mkangia committed:
Merge master: resolve conflict

mkangia committed:
add filter for source of beneficiaries

mkangia committed:
delete all migrations to make a single one

mkangia committed:
update form for better naming of variables

mkangia committed:
just retain a single model for RCH Records and keep details in Couch Doc

mkangia committed:
update views to reflect new changes

mkangia committed:
update management command to create mappings for new django version

mkangia committed:
default to show some rch records

mkangia committed:
villcode -> village_code, Village_name -> village_name

mkangia committed:
remove redundant variable

mkangia committed:
update celery tasks to pull data according to new model structure

mkangia committed:
fix dimagimon complaint

mkangia committed:
fix issue with doc type assigned incorrectly and add missing _get_rch_id_key

mkangia committed:
use get for Village_Name in case of missing attr

mkangia committed:
update place for RCH_CREDENTIALS to avoid override on localsettings

mkangia committed:
Add view for ICDS-CAS records

mkangia committed:
Add management command for import cas data

mkangia committed:
add management commands to import data [skip ci]

mkangia committed:
merge master

mkangia committed:
remove merge conflict issues

mkangia committed:
Add RCH-CAS dashboard to list of custom reports

mkangia committed:
refine filtering two combine awcid and village id properly through area mapping

mkangia committed:
add necessary decorator for the view

mkangia committed:
fix issue with filtering overriding awc_ids and mask aadhar number

mkangia committed:
fix url name for Beneficiaries list as its hooked onto icds_reports and remove the redundant url

mkangia committed:
save rch record details received as a dict on sql record itself instead of couch doc

mkangia committed:
order details in view for better readability

mkangia committed:
fix issues due to moving all properties to PG column instead of couch doc

mkangia committed:
Update the common beneficiary view to display records correctly

mkangia committed:
remove temp command file added for data import

mkangia committed:
remove temp command task added for data import

mkangia committed:
remove temp command file added for data matching exercise

mkangia committed:
squash migrations

mkangia committed:
clean up unused RCHRecordDetails and RCH_Primary_key, unused imports, help comments and lint

mkangia committed:
set choices 1,2 instead of 0,1 to keep it same as that from RCH to avoid maintaining any mapping
DATA_TYPE -> RECORD_TYPE

mkangia committed:
Disable state and district since they have just single choice as of now,
Add ability to include matched beneficiaries as well when looking at records for a single project
Put village over AWC to keep in sync with the hierarchy

mkangia committed:
revisit filtering

1. update logic for filtering to rely just on village id for RCH and just on AWC for ICDS-CAS
2. include ability to include matched records when looking for project specific records
3. clean up and lots of comments

mkangia committed:
re-enable fields to make them available as params

mkangia committed:
correct field name to dtcode

mkangia committed:
Add toggle restricted access to views

mkangia committed:
Include within HQ template, better fields and use select2 for easy choice selection

mkangia committed:
remove unused variable

mkangia committed:
Merge master

mkangia committed:
Add field mapping between CAS and RCH, and show corresponding values when viewing a beneficiary

mkangia committed:
fix error with iterator

mkangia committed:
Merge Master: Resolve conflict

mkangia committed:
update path for imports

mkangia committed:
fix error in field, state_id -> stcode

mkangia committed:
use the new location for base_section

mkangia committed:
add pagination to all listing pages

mkangia committed:
Add loader when fetching beneficiaries

mkangia committed:
extract const, mask aadhar num, order records by last modified and add edit link to name

mkangia committed:
lint

mkangia committed:
extract and trim down templates

mkangia committed:
add some documentation and remove redundant code

mkangia committed:
skip asking for two dates when they would be same anyway

mkangia committed:
mask aadhar number and add link to case detail page for ICDS-CAS records

mkangia committed:
remove duplicate view for common in favour of view under rch

mkangia committed:
inherit from BaseDomainView to have the domain default page look

mkangia committed:
let user know in case the server is not able to server request

mkangia committed:
lint

mkangia committed:
rescue from Programming Error for migrations to run

mkangia committed:
Return empty value instead of returning None to avoid errors further

mkangia committed:
add matching algorithm for a rch record

mkangia committed:
Skip setting choices unnecessary in tests

mkangia commented:
@snopoke this is ready now. 
Just need to make a couple of design improvements.

snopoke commented on custom/rch/forms.py:
```diff
@@ -0,0 +1,111 @@
+    :return: tuple of tuples
+    """
+    if settings.UNIT_TESTING:
+        return tuple()
```
This doesn't seem right. I don't see anything in this method that would prevent it from working in unit tests.

mkangia replied:
The error here https://travis-ci.org/dimagi/commcare-hq/builds/279301373

mkangia replied:
sorry, the correct build for the error is https://travis-ci.org/dimagi/commcare-hq/jobs/279305659

snopoke commented on custom/rch/forms.py:
```diff
@@ -0,0 +1,111 @@
+    stcode = forms.ChoiceField(
+        label=_("State"),
+        required=False,
+        choices=get_choices_for('stcode', 'stname'),
```
Rather than doing this statically you should do it in the form's `__init__`: https://stackoverflow.com/a/3420588/632517

That will also remove the need to catch `ProgrammingError` in the function

snopoke replied:
Doing it in `__init__` would also mean you could get all the options from one query rather than multiple queries per field

mkangia replied:
Doing this in `__init__` would mean that those queries would run each time the form is rendered. Should not we just call it once since it won't change anyway?

> Doing it in __init__ would also mean you could get all the options from one query rather than multiple queries per field

How is that?

mkangia replied:
> You could cache the query result to avoid running the query every time. This would also mean that you could update the values without having to restart the service.

These values are not expected to change very often. It would be after months it would change so there is no need to worry about the need to refresh it often.

I still think loading them just once on load is still a better call instead of moving to `__init__`. The programming error will just happen once during first migration and after that it would just return results(blank in case of local environment). I would rather have 4 queries fired once on load then have them being called each time on form load or checking for cache in case we cache it.
If you still think it would be better, happy to move it to `__init__`

esoergel replied:
Having it this way means a db query on import, which I think means that that can happen whenever we run `./manage.py anything`.  If you move it to the `__init__` method, you can slap a `@quickcache` decorator on it to cache the results across all workers, or if you prefer, a `@memoized` decorator, which will store it in memory similar to how it's done now.  In either case, it'd be lazily evaluated, so it wouldn't impact startup times.

snopoke commented on custom/rch/models.py:
```diff
@@ -0,0 +1,159 @@
+)
+
+
+class RCHRecord(models.Model):
```
Seems like there is quite a lot of forking between `mother` and `child`. Did you consider using separate models or [model inheritance](https://docs.djangoproject.com/en/1.11/topics/db/models/#model-inheritance)?

mkangia replied:
I started with that path but then ended up merging them as a single entity since that made it easier. Reverting it back to two different models.

snopoke commented on custom/rch/models.py:
```diff
@@ -0,0 +1,159 @@
+    name = models.CharField(null=True, max_length=255)
+    aadhar_num = models.BigIntegerField(null=True)
+    dob = models.DateTimeField(null=True)
+    rch_id = models.BigIntegerField(null=True)
```
should this be unique?

snopoke replied:
Also it looks like you're filtering on it so you'll almost certainly want an index

mkangia replied:
Yes.
Yes.

snopoke commented on custom/rch/models.py:
```diff
@@ -0,0 +1,159 @@
+            self.save()
+
+
+class AreaMapping(models.Model):
```
might need some indexes on some of these fields depending on how big the table will be

mkangia replied:
As of now this table has 873 rows but would grow. Adding indexes.

snopoke commented on custom/rch/utils.py:
```diff
@@ -0,0 +1,83 @@
+    return _initiate_client().service.DS_Data
+
+
+def _fetch_beneficiaries(for_date, state_id, beneficiary_type, district_id):
```
some docstrings on these functions would be helpful to know what data they return

snopoke reviewed (commented):
Haven't completed my review yet

mkangia committed:
remove redundant null false settings and add indexes for searchable fields

mkangia committed:
add documentation

mkangia committed:
add check to ensure rch id is unique across and also add index on it

mkangia committed:
make two separate models for child and mother

mkangia committed:
merge migrations

safe to delete old ones and keep just one since this has not been run anywhere else

mkangia committed:
lint

snopoke commented on custom/rch/views.py:
```diff
@@ -0,0 +1,335 @@
+
+        village_code = in_data.get('village_code')
+        if village_code:
+            village_ids = village_ids + [village_code]
```
`village_ids` will always be an empty list here so it doesn't seem useful to include it. Just have `village_ids = [village_code]`

snopoke commented on custom/rch/views.py:
```diff
@@ -0,0 +1,335 @@
+        # cas records by their awc_ids set as owner_id
+        village_code = in_data.get('village_code')
+        if village_code:
+            awc_ids = awc_ids + AreaMapping.fetch_awc_ids_for_village_id(village_code)
```
same here re `awc_ids` which will always be empty to no need to include it

snopoke commented on custom/rch/views.py:
```diff
@@ -0,0 +1,335 @@
+        include_matched_beneficiaries = (in_data.get('matched') == 'on')
+        size = in_data.get('limit')
+        offset = (in_data.get('page') - 1) * in_data.get('limit')
+        if self._set_cas_beneficiaries(include_matched_beneficiaries, in_data=in_data):
```
this should rather just return the beneficiaries rather than setting them as an attribute of the class:
```
beneficiaries = self._get_cas_beneficiaries(include_matched_beneficiaries, in_data=in_data)
if beneficiaries:
   ....
```

snopoke commented on custom/rch/views.py:
```diff
@@ -0,0 +1,335 @@
+        offset = (in_data.get('page') - 1) * in_data.get('limit')
+        if self._set_cas_beneficiaries(include_matched_beneficiaries, in_data=in_data):
+            cas_records = self.beneficiaries.start(offset).size(size).run()
+            if cas_records:
```
I think this will always be True. If you want to test to see if there are results you should use `cas_records.total`

snopoke commented on custom/rch/views.py:
```diff
@@ -0,0 +1,335 @@
+                context['beneficiaries_count'] = len(cas_records.hits)
+
+    def _set_context_for_rch_records(self, context, in_data=None):
+        include_matched_beneficiaries = (in_data.get('matched') == 'on')
```
nit: unnecessary brackets. Also in other places

mkangia replied:
just added intentionally for better readability.

snopoke commented on custom/rch/views.py:
```diff
@@ -0,0 +1,335 @@
+            # ToDo: remove awc_ids added for demo purpose
+            user_ids = user_ids + awc_loc_ids + awc_ids
+            if user_ids:
+                self.beneficiaries = self.beneficiaries.filter(case_es.user(user_ids))
```
you can just go `beneficiaries.user(user_ids)`

snopoke commented on custom/rch/views.py:
```diff
@@ -0,0 +1,335 @@
+        if in_data.get('present_in') == 'cas':
+            itemList = context.get('beneficiaries')
+            for item in itemList:
+                item['editUrl'] = reverse(CaseDetailsView.urlname, args=(ICDS_CAS_DOMAIN, item.get('_id')))
```
can this not be done in JS? Doesn't seem great to do essentially the same reverse for every item.
```
# in python
context['case_details_view_url'] = reverse(CaseDetailsView.urlname, args=[request.domain, '---'])

# in template
{% initial_page_data 'case_details_view_url' case_details_view_url %}

# in JS
var url = hqImport('hqwebapp/js/initial_page_data').reverse('case_details_view_url', item._id)
```

snopoke replied:
Same in `_format_beneficiary`

snopoke commented on custom/rch/views.py:
```diff
@@ -0,0 +1,335 @@
+    def page_url(self):
+        return reverse(self.urlname, args=[ICDS_CAS_DOMAIN, self.beneficiary_type, self.beneficiary_id])
+
+    @method_decorator(require_superuser)
```
move to class level

snopoke commented on settings.py:
```diff
@@ -2205,3 +2208,23 @@ def _determine_couch_databases(couch_databases):
+RCH_PERMITTED_FIELDS = {
+    'mother': extract_rch_fields_from_mapping('mother'),
+    'child': extract_rch_fields_from_mapping('child'),
+}
```
would be nice to make these lazy so that they don't slow down the startup of every Django process

mkangia commented on custom/rch/models.py:
```diff
@@ -0,0 +1,156 @@
+
+    @property
+    def rch_id_key(self):
+        return self._rch_id_key
```
I prefer to use this when i want to have a private property that should not be updated and just expose a method to make it available for read. Probably safe to use `self._rch_id_key` directly if used within class properties but using `self. rch_id_key()` when accessed outside of class.

mkangia commented on custom/rch/utils.py:
```diff
@@ -0,0 +1,87 @@
+        data = response['_value_1'].find('NewDataSet')
+        response_content = etree_to_dict(data)['NewDataSet']
+        if (response_content[0].get('Table1', None) and
+                response_content[0]['Table1'][0]['Message'] == 'Please check your URL or Security Code'):
```
the response code is always 200 so need to rely on the message.
making it a constant.

mkangia commented on custom/rch/views.py:
```diff
@@ -0,0 +1,335 @@
+
+    @method_decorator(require_superuser)
+    @use_select2
+    @use_angular_js
```
Need to set the cookie as well. So `dispatch` would need to stay. 
You probably added this review before i added the cookie set up.

mkangia committed:
don't store beneficiaries on self and just return it instead plus other improvements

mkangia committed:
move rch app constants under rch app

mkangia commented:
:(
